tools/gfx/metal/metal-device.cpp

```diff
@@ -212,107 +288,176 @@ Result DeviceImpl::createTextureResource(
+    }
+    if (desc.allowedStates.contains(ResourceState::UnorderedAccess))
+    {
+        textureUsage |= MTL::TextureUsageShaderWrite;
+        textureUsage |= MTL::TextureUsageShaderAtomic;
+    }
+
+    textureDesc->setMipmapLevelCount(desc.numMipLevels);
+    textureDesc->setArrayLength(arrayLength);
+    textureDesc->setPixelFormat(pixelFormat);
+    textureDesc->setUsage(textureUsage);
+    textureDesc->setSampleCount(desc.sampleDesc.numSamples);
+    textureDesc->setAllowGPUOptimizedContents(desc.memoryType == MemoryType::DeviceLocal);
+
+    textureImpl->m_texture = NS::TransferPtr(m_device->newTexture(textureDesc.get()));
+    if (!textureImpl->m_texture)
+    {
+        return SLANG_FAIL;
+    }
+
+    returnComPtr(outResource, textureImpl);
     return SLANG_OK;
 }
 
 Result DeviceImpl::createBufferResource(
     const IBufferResource::Desc& descIn, const void* initData, IBufferResource** outResource)
 {
+    AUTORELEASEPOOL
+
     BufferResource::Desc desc = fixupBufferDesc(descIn);
 
     const Size bufferSize = desc.sizeInBytes;
 
-    MTL::ResourceOptions opts = (desc.memoryType == MemoryType::DeviceLocal ? MTL::ResourceStorageModePrivate : 0);
-
-    RefPtr<BufferResourceImpl> buffer(new BufferResourceImpl(desc, this));
+    MTL::ResourceOptions resourceOptions = MTL::ResourceOptions(0);
+    switch (desc.memoryType)
+    {
+    case MemoryType::DeviceLocal:
+        resourceOptions = MTL::ResourceStorageModePrivate;
+        break;
+    case MemoryType::Upload:
+        resourceOptions = MTL::ResourceStorageModeShared | MTL::CPUCacheModeWriteCombined;
+        break;
+    case MemoryType::ReadBack:
+        resourceOptions = MTL::ResourceStorageModeShared;
+        break;
+    }
+    resourceOptions |= (desc.memoryType == MemoryType::DeviceLocal) ? MTL::ResourceStorageModePrivate : MTL::ResourceStorageModeShared;
 
-    if (initData)
+    RefPtr<BufferResourceImpl> bufferImpl(new BufferResourceImpl(desc, this));
+    bufferImpl->m_buffer = NS::TransferPtr(m_device->newBuffer(bufferSize, resourceOptions));
+    if (!bufferImpl->m_buffer)
     {
-        buffer->m_buffer = m_device->newBuffer(initData, bufferSize, opts);
+        return SLANG_FAIL;
     }
-    else
+
+    if (initData)
     {
-        buffer->m_buffer = m_device->newBuffer(bufferSize, opts);
+        NS::SharedPtr<MTL::Buffer> stagingBuffer = NS::TransferPtr(m_device->newBuffer(
+            initData, bufferSize, MTL::ResourceStorageModeShared | MTL::CPUCacheModeWriteCombined));
+        MTL::CommandBuffer* commandBuffer = m_commandQueue->commandBuffer();
+        MTL::BlitCommandEncoder* encoder = commandBuffer->blitCommandEncoder();
+        if (!stagingBuffer || !commandBuffer || !encoder)
+        {
+            return SLANG_FAIL;
+        }
+        encoder->copyFromBuffer(stagingBuffer.get(), 0, bufferImpl->m_buffer.get(), 0, bufferSize);
+        encoder->endEncoding();
+        commandBuffer->commit();
```

Comment:
In cases where the resource is used in a subsequent command buffer submission this should be fine, as command buffers are executed in the order they are submitted. However, there might be an edge case where we upload to a host visible buffer and directly after want to read/modify the host mapped memory. I guess there are two solutions:
- wait for the submission
- for host visible buffers, write to the mapped memory instead of issuing a copy command
For now, I think I'll just add the wait.